core/src/main/java/org/hisp/dhis/android/core/relationship/internal/RelationshipImportHandler.kt

```diff
@@ -49,23 +51,61 @@ internal class RelationshipImportHandler internal constructor(
     ) {
         importSummaries?.filterNotNull()?.forEach { importSummary ->
             importSummary.reference()?.let { relationshipUid ->
-                val relationship = relationshipRepository.withItems().uid(relationshipUid).blockingGet()
+                handleSingleRelationship(importSummary, relationshipUid)
+            }
+        }
 
-                val state = getSyncState(importSummary.status())
+        processIgnoredRelationships(importSummaries, relationships)
+    }
 
-                val handledState =
-                    if (state == State.ERROR || state == State.WARNING) {
-                        State.TO_UPDATE
-                    } else {
-                        state
-                    }
+    private suspend fun handleSingleRelationship(
+        importSummary: RelationshipImportSummary,
+        relationshipUid: String,
+    ) {
+        val relationship = relationshipRepository.withItems().uid(relationshipUid).blockingGet()
+        val relationshipNotFoundOnServer = checkRelationshipNotFoundOnServer(importSummary)
 
-                relationshipStore.setSyncStateOrDelete(relationshipUid, handledState)
-                dataStatePropagator.propagateRelationshipUpdate(relationship)
-            }
+        val handleAction = if (relationshipNotFoundOnServer) {
+            handleRelationshipNotFound(relationship)
+        } else {
+            handleRelationshipUpdateState(importSummary, relationshipUid)
         }
 
-        processIgnoredRelationships(importSummaries, relationships)
+        if (handleAction != HandleAction.Delete) {
+            dataStatePropagator.propagateRelationshipUpdate(relationship)
+        }
+    }
+
+    private suspend fun handleRelationshipNotFound(relationship: Relationship?): HandleAction {
+        relationship?.let { relationshipStore.deleteByEntity(it) }
+        return HandleAction.Delete
+    }
+
+    private suspend fun handleRelationshipUpdateState(
+        importSummary: RelationshipImportSummary,
+        relationshipUid: String,
+    ): HandleAction {
+        val state = getSyncState(importSummary.status())
+        val handledState = if (state == State.ERROR || state == State.WARNING) {
+            State.TO_UPDATE
+        } else {
+            state
+        }
+
+        relationshipStore.setSyncStateOrDelete(relationshipUid, handledState)
```

Comment:
I think this line can return a `HandleAction.Delete` if the relationship was successfully deleted. Shouldn't this whole method return that action?